contributors/design-proposals/api-machinery/customresource-conversion-webhook.md

```diff
@@ -0,0 +1,861 @@
+  // Optional, correspond to the first version in the versions list 
+  Subresources *CustomResourceSubresources
+  Versions []CustomResourceDefinitionVersion
+  // Optional, and correspond to the first version in the versions list  
+  AdditionalPrinterColumns []CustomResourceColumnDefinition
+
+  Conversion *CustomResourceConversion
+}
+
+type CustomResourceDefinitionVersion struct {
+  Name string
+  Served Boolean
+  Storage Boolean
+  // These three fields should not be set for first item in Versions list
+  Schema *JSONSchemaProp
+  Subresources *CustomResourceSubresources
+  AdditionalPrinterColumns []CustomResourceColumnDefinition
+}
+
+Type CustomResourceConversion struct {
+  // Conversion strategy, either "nop” or "webhook”. If webhook is set, Webhook field is required.
+  Strategy string
+
+  // Additional information for external conversion if strategy is set to external
+  // +optional
+  Webhook *CustomResourceConversionWebhook
+}
+
+type CustomResourceConversionWebhook {
+  // ClientConfig defines how to communicate with the webhook. This is the same config used for validating/mutating webhooks.
+  ClientConfig WebhookClientConfig
+}
+```
+
+## Defaulting
+
+In case that there is no versions list, a single version with values defaulted to top level version will be created. That means a single version with a name set to spec.version.
+
+
+## Validation
+
+To keep backward compatibility, the top level fields (schema, additionalPrinterColumns or subresources) are stay the same and source of truth for first (top) version. The first item in the versions list should not set any of those fields. The plan is to use unified version list for v1.
+
+
+## Support Level
+
+The feature will be alpha in the first supported version and will have a feature gate that is defaulted to false. The roll-back story with a feature gate is much more clear. if we have the features as alpha in version Y (>X where the feature is missing) and we make it beta in version Z, it is not safe to use the feature and downgrade from Y to X but the feature is alpha in Y which is fine. It is safe to downgrade from Z to Y (given that we enable the feature gate in Y) and that is desirable as the feature is beta in Z. I will add this detail to the doc.
+
+
+## Rollback
+
+Users that need to rollback to version X (but may currently be running version Y > X) of apiserver should not use CRD Webhook Conversion until X is a version that supports these features.  If a user were to create a CRD that uses CRD Webhook Conversion and then rolls back then the following would happen:
+
+1. The stored custom resources in etcd will not be deleted.
+
+2. Any clients that try to get the custom resources will get a 500 (internal server error). this is distinguishable from a deleted object for get and the list operation will also fail. That means the CRD is not served at all and Clients that try to garbage collect related resources to missing CRs should be aware of this. 
```

Comment:
Why does this happen? X must at least support webhook versioning as alpha. Then we can throw 500 errors if the gate is disabled.

Reply:
you mean CRD versioning with alpha webhook conversion? The assumption of this section is version X does not support webhook conversion and we do have a webhook conversion. No support for webhook conversion means the feature gate is off.